address.go

```diff
@@ -81,6 +118,37 @@ type Address interface {
 // When the address does not encode the network, such as in the case of a raw
 // public key, the address will be associated with the passed defaultNet.
 func DecodeAddress(addr string, defaultNet *chaincfg.Params) (Address, error) {
+
+	// Bech32 encoded segwit addresses start with an human-readable part
+	// (hrp) followed by 1. For Bitcoin mainnet the hrp is "bc", and for
+	// testnet it is "tb". If the address string matches the hrp found in
+	// the defaultNet params, we try to decode it as a segwit address.
+	hrp := defaultNet.Bech32HRPSegwit
+	prefix := hrp + "1"
+	if strings.HasPrefix(strings.ToLower(addr), prefix) {
+		witnessVer, witnessProg, err := decodeSegWitAddress(hrp, addr)
+		if err != nil {
+			return nil, err
+		}
+
+		// We currently only support P2WPKH and P2WSH, which is
+		// witness version 0.
+		if witnessVer != 0 {
+			return nil, fmt.Errorf("unsupported witness "+
```

Comment:
Please define programmatically detectable errors for these as the rest of the function does.

```Go
// Comments...
type UnsupportedWitnessVerError byte
func (e UnsupportedWitnessVerError) Error() string {
	return "unsupported witness version: " + string(e)
}


...
...
	if witnessVer != 0 {
		return nil, UnsupportedWitnessVerError(witnessVer)
	}
```